### PR TITLE
Dmitri/k8s 1.7

### PR DESCRIPTION
**Pull request by a-palchikov**

### CODE REVIEW

**Comment by klizhentas** [comment]
can you switch to golang/dep and also use latest trace to use lowercase `sirupsen` here as well

**Comment by a-palchikov** [comment]
Ok, will do

**Comment by a-palchikov** [comment]
After a bit of struggle with gravity I want to postpone updating to lower-case sirupsen to another PR and just include the k8s 1.7-related changes here.

File: deployment.go [hunk 1]
```diff
@@ -82,21 +85,45 @@ func (c *DeploymentControl) Delete(ctx context.Context, cascade bool) error {
...
+
+	errDelete := waitForObjectDeletion(func() error {
+		_, err := deployments.Get(c.deployment.Name, metav1.GetOptions{})
+		return ConvertError(err)
+	})
+
+	// wait until all Pods have been cleaned up
+	err = waitForPods(pods, currentPods, *c.Entry)
+	if err == nil {
+		err = errDelete
```
Comment (r0mant) [inline deployment.go]: Can you explain this logic? Why do we not return errDelete above and assign it here instead?

File: utils.go [hunk 2]
```diff
@@ -281,3 +283,83 @@ func isEmptyDetails(details *unversioned.StatusDetails) bool {
...
+
+func waitForObjectDeletion(fn func() error) error {
+	return wait.PollImmediate(deletePollInterval, deleteTimeout, func() (bool, error) {
+		switch err := fn(); {
+		case err == nil:
+			return false, nil
+		case trace.IsNotFound(err):
+			return true, nil
+		default:
+			return false, err
```
Comment (r0mant) [inline utils.go]: trace.Wrap(err)

File: utils.go [hunk 3]
```diff
@@ -281,3 +283,83 @@ func isEmptyDetails(details *unversioned.StatusDetails) bool {
...
+
+func deletePodsList(podIface corev1.PodInterface, pods []v1.Pod, entry log.Entry) error {
+	for _, pod := range pods {
+		entry.Debugf("deleting pod %v", pod.Name)
+		err := ConvertError(podIface.Delete(pod.Name, nil))
+		if err != nil && !trace.IsNotFound(err) {
+			return ConvertError(err)
+		}
+	}
+	return nil
```
Comment (r0mant) [inline utils.go]: Is it on purpose that this method doesn't "waitForPods" and the one below does? Also, do we need two similar methods? This one is used in a single place and you could convert slice to a map there and reuse the one below.

Comment (a-palchikov) [reply utils.go]: I thought it was not worth the trouble - this code is in bad need for a rewrite that would provide more flexibility when adding new types (it is a lot of work currently), so I'd rather have a good stab at restructuring this for a greater good at a later point.

Comment (r0mant) [reply utils.go]: :+1:

File: deployment.go (hunk 1 above)
Comment (a-palchikov) [reply deployment.go]: you're right - we could return an error sooner.

**Pull request merged**